### PR TITLE
Bug 100627 - Re-Packaging Phase 2: junixsocket

### PR DESCRIPTION
**Pull request by msmcgillis**
Probably want to rework the /opt/zimbra/jetty distribution in this once a better solution is worked out.

If we can narrow down exactly which specific jars are required we can reduce the number of jars currently getting pulled into this.

I'm not convinced everything is correct for the debug packages generated. It isn't clear to me that the nar files will be right in those.

But this should be a good start that can be tested with to find better answers to the above.

### CODE REVIEW

**Comment by quanah** [comment]
This will need to build-depend on the zimbra-maven package.  It will need to install and remove zimbra-maven.  It needs to correctly export the various maven environment variables so that maven doesn't download a ton of crap into ~build/.m2/.  See the zimbra-maven package for details.

**Comment by quanah** [comment]
zimbra-maven package is available to install on your vagrant VMs.  All ZCS packages should depend on zimbra-base as well, which is missing in the package definitions here as well.

**Comment by msmcgillis** [comment]
See my above commits for getting a little closer.

In theory that might work but currently it does not for ubuntu. Maven tries to download things from HTTPS servers to do this it needs a java truststore file the current default from the zimbra-java that is getting installed and used is a symbolic link to  something that doesn't exist.

```
$ ls -al /opt/zimbra/common/lib/jvm/openjdk-1.8.0_92-zimbra/jre/lib/security/cacerts
lrwxrwxrwx 1 root root 34 Apr 29 15:30 /opt/zimbra/common/lib/jvm/openjdk-1.8.0_92-zimbra/jre/lib/security/cacerts -> ../../../../../../etc/java/cacerts
```

So the build fails with SSL communication errors during download.

The maven errors look something like this:

```
$ make
mkdir -p /home/build/git/90/packages/thirdparty/junixsocket/build/UBUNTU14_64/src
(cd /home/build/git/90/packages/thirdparty/junixsocket/build/UBUNTU14_64/src && \
	git clone https://github.com/kohlschutter/junixsocket.git junixsocket >/dev/null 2>&1 && \
	cd junixsocket && \
        git reset --hard junixsocket-parent-2.0.4 >/dev/null 2>&1 && \
        rm -rf .git .gitignore && \
	cd .. && \
	mv junixsocket junixsocket-2.0.4 && \
	tar cfz junixsocket-2.0.4.tgz junixsocket-2.0.4 && \
	rm -rf junixsocket-2.0.4 \
	)
sudo apt-get --assume-yes --verbose-versions install zimbra-base zimbra-maven
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following extra packages will be installed:
   zimbra-openjdk (1.8.0u92b14-1zimbra9.0b3.14.04)
The following NEW packages will be installed:
   zimbra-base (1.0.1-1zimbra9.0b1.14.04)
   zimbra-maven (3.3.9-1zimbra9.0b1.14.04)
   zimbra-openjdk (1.8.0u92b14-1zimbra9.0b3.14.04)
0 upgraded, 3 newly installed, 0 to remove and 21 not upgraded.
Need to get 0 B/202 MB of archives.
After this operation, 318 MB of additional disk space will be used.
Selecting previously unselected package zimbra-base.
(Reading database ... 33225 files and directories currently installed.)
Preparing to unpack .../zimbra-base_1.0.1-1zimbra9.0b1.14.04_all.deb ...
Unpacking zimbra-base (1.0.1-1zimbra9.0b1.14.04) ...
Selecting previously unselected package zimbra-openjdk.
Preparing to unpack .../zimbra-openjdk_1.8.0u92b14-1zimbra9.0b3.14.04_amd64.deb ...
Unpacking zimbra-openjdk (1.8.0u92b14-1zimbra9.0b3.14.04) ...
Selecting previously unselected package zimbra-maven.
Preparing to unpack .../zimbra-maven_3.3.9-1zimbra9.0b1.14.04_amd64.deb ...
Unpacking zimbra-maven (3.3.9-1zimbra9.0b1.14.04) ...
Setting up zimbra-base (1.0.1-1zimbra9.0b1.14.04) ...
usermod: no changes
Setting up zimbra-openjdk (1.8.0u92b14-1zimbra9.0b3.14.04) ...
Setting up zimbra-maven (3.3.9-1zimbra9.0b1.14.04) ...
mkdir -p /home/build/git/90/packages/thirdparty/junixsocket/build/UBUNTU14_64
cp -prf zimbra-* /home/build/git/90/packages/thirdparty/junixsocket/build/UBUNTU14_64
cd /home/build/git/90/packages/thirdparty/junixsocket/build/UBUNTU14_64/zimbra-junixsocket && \
	  perl -pi -e 's|VERSION|2.0.4|; s|ITERATION|1zimbra9.0b1|; s|ZAPPEND|.14.04|' debian/changelog  && \
	  perl -pi -e 's|OZCLE|/opt/zimbra/common/libexec|g; s|OZCB|/opt/zimbra/common/bin|g; s|OZCE|/opt/zimbra/common/etc|g; s|OZCI|/opt/zimbra/common/include|g; s|OZCL|/opt/zimbra/common/lib|g; s|OZCS|/opt/zimbra/common/share|g; s|OZC|/opt/zimbra/common|g;' debian/rules && \
	cp /home/build/git/90/packages/thirdparty/junixsocket/build/UBUNTU14_64/src/junixsocket-2.0.4.tgz ../zimbra-junixsocket_2.0.4.orig.tar.gz && \
	tar xfz ../zimbra-junixsocket_2.0.4.orig.tar.gz --strip-components=1 && \
	dpkg-buildpackage -irpm
dpkg-buildpackage: source package zimbra-junixsocket
dpkg-buildpackage: source version 2.0.4-1zimbra9.0b1
dpkg-buildpackage: source distribution unstable
dpkg-buildpackage: source changed by Zimbra Packaging Services <packaging-devel@zimbra.com>
dpkg-buildpackage: host architecture amd64
 dpkg-source -irpm --before-build zimbra-junixsocket
 fakeroot debian/rules clean
make[1]: Entering directory `/home/build/git/90/packages/thirdparty/junixsocket/build/UBUNTU14_64/zimbra-junixsocket'
dh clean
   dh_testdir
   dh_auto_clean
   dh_clean
make[1]: Leaving directory `/home/build/git/90/packages/thirdparty/junixsocket/build/UBUNTU14_64/zimbra-junixsocket'
 dpkg-source -irpm -b zimbra-junixsocket
dpkg-source: info: using source format `3.0 (quilt)'
dpkg-source: info: building zimbra-junixsocket using existing ./zimbra-junixsocket_2.0.4.orig.tar.gz
dpkg-source: info: building zimbra-junixsocket in zimbra-junixsocket_2.0.4-1zimbra9.0b1.debian.tar.gz
dpkg-source: info: building zimbra-junixsocket in zimbra-junixsocket_2.0.4-1zimbra9.0b1.dsc
 debian/rules build
make[1]: Entering directory `/home/build/git/90/packages/thirdparty/junixsocket/build/UBUNTU14_64/zimbra-junixsocket'
dh build
   dh_testdir
   debian/rules override_dh_auto_build
make[2]: Entering directory `/home/build/git/90/packages/thirdparty/junixsocket/build/UBUNTU14_64/zimbra-junixsocket'
mvn -Dmaven.repo.local=/home/build/git/90/packages/thirdparty/junixsocket/build/UBUNTU14_64/zimbra-junixsocket/.m2/repository -P with-native package
[INFO] Scanning for projects...
Downloading: https://repo.maven.apache.org/maven2/com/kohlschutter/kohlschutter-parent/1.1/kohlschutter-parent-1.1.pom
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[FATAL] Non-resolvable parent POM for com.kohlschutter.junixsocket:junixsocket-parent:2.0.4: Could not transfer artifact com.kohlschutter:kohlschutter-parent:pom:1.1 from/to central (https://repo.maven.apache.org/maven2): java.lang.RuntimeException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty and 'parent.relativePath' points at wrong local POM @ line 8, column 11
 @ 
[ERROR] The build could not read 1 project -> [Help 1]
[ERROR]   
[ERROR]   The project com.kohlschutter.junixsocket:junixsocket-parent:2.0.4 (/home/build/git/90/packages/thirdparty/junixsocket/build/UBUNTU14_64/zimbra-junixsocket/pom.xml) has 1 error
[ERROR]     Non-resolvable parent POM for com.kohlschutter.junixsocket:junixsocket-parent:2.0.4: Could not transfer artifact com.kohlschutter:kohlschutter-parent:pom:1.1 from/to central (https://repo.maven.apache.org/maven2): java.lang.RuntimeException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty and 'parent.relativePath' points at wrong local POM @ line 8, column 11 -> [Help 2]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/ProjectBuildingException
[ERROR] [Help 2] http://cwiki.apache.org/confluence/display/MAVEN/UnresolvableModelException
make[2]: *** [override_dh_auto_build] Error 1
make[2]: Leaving directory `/home/build/git/90/packages/thirdparty/junixsocket/build/UBUNTU14_64/zimbra-junixsocket'
make[1]: *** [build] Error 2
make[1]: Leaving directory `/home/build/git/90/packages/thirdparty/junixsocket/build/UBUNTU14_64/zimbra-junixsocket'
dpkg-buildpackage: error: debian/rules build gave error exit status 2
make: *** [build_deb] Error 2
```

**Comment by msmcgillis** [comment]
In that latest code for .m2 I use:

rpm
mvn -Dmaven.repo.local=%{buildroot}/.m2/repository -P with-native package

deb
mvn -Dmaven.repo.local=${CURDIR}/.m2/repository -P with-native package

It might be better if we can find a way to bake the above into the actual zimbra-maven packages settings.xml file in the maven conf directory. Then we have just one place to adjust this as we move forward but if you prefer the above it should work once we resolve the above error.

**Comment by quanah** [comment]
You can install zimbra-openjdk-cacerts to resolve the CA certs issue.

**Comment by msmcgillis** [comment]
I added zimbra-openjdk-cacerts as a test and it doesn't by it self solve it.

First it becomes a little odd from the junixsocket perspective. junixsocket needs zimbra-maven it is this package that on ubuntu decides it needs openjdk so I think it probably would be better to have it decide to also always bring in zimbra-openjdk-cacerts. Otherwise I ether need to customes junixsocket to do custom things on the package side based on the OS or as result of adding zimbra-openjdk-cacerts to zimbra-junixsocket will result in pulling in a bunch of things on the CentOS side that may not even exist or if it does is not really needed/wanted.

Outside of the above the build still fails with the same error. I can force it to work with this:

```
mvn -Dmaven.repo.local=/home/build/git/90/packages/thirdparty/junixsocket/build/UBUNTU14_64/zimbra-junixsocket/.m2/repository -Djavax.net.ssl.trustStore=/opt/zimbra/common/lib/jvm/openjdk-1.8.0_92-zimbra/jre/lib/security/cacerts -Djavax.net.ssl.trustStorePassword=changeit -D with-native package 
```

But I would think the java package or the cacert one  would take care of the above for me on default or even the maven one but I think java will usually deal with this.

**Comment by quanah** [comment]
I'm confused when you say that ubuntu decides it needs openjdk.  It sounds to me like you are continuing to use the OS provided maven and not our zimbra-maven build.  I would expect the call to maven to be /opt/zimbra/common/bin/mvn rather than just "mvn".

**Comment by quanah** [comment]
Also, I see a lack of configuring JAVA_HOME env variable, which is likely going to be necessary in the debian/rules and rpm/SPEC files.

**Comment by msmcgillis** [comment]
The above code now builds on both OS's using the vagrant instances on genesis that where created for me.

In the latest code on the CentOS side I added to the makefile permission cleanup that probably should exist in the zimbra-openjdk package.  After I install the packages I'm currently doing:

chmod 644 /opt/zimbra/common/lib/jvm/openjdk-1.8.0_92-zimbra/lib/ct.sym

It is getting installed as 600 with root as owner so my code can't access the file during builds and will blow up with out this. If this gets fixed upstream let me know and I will be happy to remove from this code.

**Comment by quanah** [comment]
That would seem to be a RHEL bit.  The perms are correct on Ubuntu based OSes.  I'll fix that in our zimbra-openjdk package.

**Comment by quanah** [comment]
Fixed in our openjdk package now.  I think that is why ant & maven were blowing up when I was attempting to compile them as well.  Now I can revisit those packages.

**Comment by msmcgillis** [comment]
The above commit removes that chmod.

The code was tested on genesis for both OS's and built without errors.

File: thirdparty/junixsocket/zimbra-junixsocket/debian/control [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+Multi-Arch: same
+Depends: ${shlibs:Depends}, ${misc:Depends}
+Description: junixsocket JARs
+
+Package: zimbra-junixsocket-dbg
+Section: debug
+Priority: extra
+Architecture: any
+Multi-Arch: same
+Depends: zimbra-junixsocket-lib (= ${binary:Version}), ${misc:Depends}
```
Comment (quanah) [inline thirdparty/junixsocket/zimbra-junixsocket/debian/control]: There is no lib package, so depending on it seems problematic?

File: thirdparty/junixsocket/zimbra-junixsocket/debian/control [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+Uploaders: Zimbra Packaging Services <packaging-devel@zimbra.com>
+Standards-Version: 3.9.5
+Homepage:  https://github.com/kohlschutter/junixsocket
+
+Package: zimbra-junixsocket
+Section: jars
+Priority: important
+Architecture: any
+Multi-Arch: same
+Depends: ${shlibs:Depends}, ${misc:Depends}
```
Comment (quanah) [inline thirdparty/junixsocket/zimbra-junixsocket/debian/control]: Needs to depend on zimbra-base as well.

File: thirdparty/junixsocket/zimbra-junixsocket/rpm/SPECS/junixsocket.spec [hunk 3]
```diff
@@ -0,0 +1,43 @@
+Summary:            Zimbra's junixsocket build
+Name:               zimbra-junixsocket
+Version:            2.0.4
+Release:            1zimbra9.0b1
+BuildRequires:      zimbra-maven
+License:            Apache-2.0
+Source:             %{name}-%{version}.tgz
+URL:                https://github.com/kohlschutter/junixsocket
+
```
Comment (quanah) [inline thirdparty/junixsocket/zimbra-junixsocket/rpm/SPECS/junixsocket.spec]: Missing a Requires: line.  Must require zimbra-base.

**Comment by msmcgillis** [comment]
Should be all set with those three items in the above commit.

File: thirdparty/junixsocket/zimbra-junixsocket/debian/control [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+Multi-Arch: same
+Depends: ${shlibs:Depends}, ${misc:Depends}, zimbra-base
+Description: junixsocket JARs
+
+Package: zimbra-junixsocket-dbg
+Section: debug
+Priority: extra
+Architecture: any
+Multi-Arch: same
+Depends: ${shlibs:Depends}, ${misc:Depends}, zimbra-base
```
Comment (quanah) [inline thirdparty/junixsocket/zimbra-junixsocket/debian/control]: the debug package should depend on the zimbra-junixsocket package, no?  Seems like wouldn't be much point to installing it w/o the base package. ;)

File: thirdparty/junixsocket/zimbra-junixsocket/rpm/SPECS/junixsocket.spec [hunk 5]
```diff
@@ -0,0 +1,44 @@
+Summary:            Zimbra's junixsocket build
+Name:               zimbra-junixsocket
+Version:            2.0.4
+Release:            1zimbra9.0b1
```
Comment (quanah) [inline thirdparty/junixsocket/zimbra-junixsocket/rpm/SPECS/junixsocket.spec]: The version and release lines are incorrect.  Please see other SPEC files for correct format.

File: thirdparty/junixsocket/zimbra-junixsocket/debian/control [hunk 6]
```diff
@@ -0,0 +1,24 @@
+Source: zimbra-junixsocket
+Build-Depends: debhelper (>= 9), m4, dpkg-dev (>= 1.15.7), maven3
```
Comment (quanah) [inline thirdparty/junixsocket/zimbra-junixsocket/debian/control]: build-depends is incorrect.  We should have no dep on maven3.  It should be zimbra-maven instead

File: thirdparty/junixsocket/zimbra-junixsocket/debian/changelog [hunk 7]
```diff
@@ -0,0 +1,5 @@
+zimbra-junixsocket (2.0.4-1zimbra9.0b1) unstable; urgency=medium
```
Comment (quanah) [inline thirdparty/junixsocket/zimbra-junixsocket/debian/changelog]: Please see other changelog files for correct format of the versioning here.

File: thirdparty/junixsocket/zimbra-junixsocket/debian/control (hunk 4 above)
Comment (msmcgillis) [inline thirdparty/junixsocket/zimbra-junixsocket/debian/control]: Hard to say. To be honest as it is on both OS's the debug packages contain either nothing or nothing useful.

If you know how to get maven to generate useful debug stuff that is compatible with these package managers I'm happy to fix. 

But just doing some basic digging through the POM files and looking over maven I'm not seeing anything that offers any clear clues.

On top of that we are providing a nar file for the libraries and not a traditional library.

So over all to be honest I'm not clear on expectations in this context.

Given all the above I would tend to lean more towards dumping them.

Let me know what direction you want to go and if you want them how you recommend generating what it is you are after.

Comment (quanah) [inline thirdparty/junixsocket/zimbra-junixsocket/debian/control]: I'm fine with dumping them if they generate nothing useful.

**Comment by msmcgillis** [comment]
Should be all set with those four items in the above commit.

**Comment by quanah** [comment]
+1

**Pull request merged**